### ISSUE TITLE
unused_allocation warning should ignore allocations to pass Box to things taking self: &Box

### ISSUE THREAD

**Issue by ssbr** [rust-lang/rust#151846]
### Code

```Rust
// Copyright 2026 Google LLC.
// SPDX-License-Identifier: Apache-2.0

struct MyStruct;

trait TraitTakesBoxRef {
    fn trait_takes_box_ref(&self);
}

impl TraitTakesBoxRef for Box<MyStruct> {
    fn trait_takes_box_ref(&self) {}
}

impl MyStruct {
    fn inherent_takes_box_ref(self: &Box<Self>) {}
}

fn takes_box_ref(_: &Box<MyStruct>) {}

trait TraitTakesBoxVal {
    fn trait_takes_box_val(self);
}

impl TraitTakesBoxVal for Box<MyStruct> {
    fn trait_takes_box_val(self) {}
}

impl MyStruct {
    fn inherent_takes_box_val(self: Box<Self>) {}
}

fn takes_box_val(_: Box<MyStruct>) {}

pub fn foo() {
    Box::new(MyStruct).trait_takes_box_ref();     // warns
    Box::new(MyStruct).inherent_takes_box_ref();  // warns
    takes_box_ref(&Box::new(MyStruct));           // OK

    Box::new(MyStruct).trait_takes_box_val();     // OK
    Box::new(MyStruct).inherent_takes_box_val();  // OK
    takes_box_val(Box::new(MyStruct));            // OK
    
    // fully-qualified syntax also does not warn:
    <Box<MyStruct> as TraitTakesBoxRef>::trait_takes_box_ref(&Box::new(MyStruct));
    MyStruct::inherent_takes_box_ref(&Box::new(MyStruct));
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
warning: unnecessary allocation, use `&` instead
  --> src/lib.rs:35:5
   |
35 |     Box::new(MyStruct).trait_takes_box_ref();     // warns
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_allocation)]` (part of `#[warn(unused)]`) on by default

warning: unnecessary allocation, use `&` instead
  --> src/lib.rs:36:5
   |
36 |     Box::new(MyStruct).inherent_takes_box_ref();  // warns
   |     ^^^^^^^^^^^^^^^^^^

warning: `playground` (lib) generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.46s
```

### Desired output

```Shell
Compiling playground v0.0.1 (/playground)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.47s
```

### Rationale and extra context

If a function accepts a `&Box`, there is no alternative: you _must_ pass a Box, and can't avoid the allocation. The warning understands this for ordinary functions that accept a parameter of type `&Box<T>` and `Box<T>`. But, for methods, it only understands that `self: Box<Self>` requires the box. It does not suppress the warning for `self: &Box<Self>` unless you call the method using fully-qualified function call syntax.

### Other cases

```Rust

```

### Rust Version

```Shell
rust playground

Stable channel
Build using the Stable version: 1.93.0
```
